In [1]:
import pandas as pd
import numpy as np

import matplotlib
import os

# 작업형 1 - 1 
### 보스턴 데이터 범죄율 컬럼 top10 중 10번째 범죄율 값으로 1~10위의 범죄율 값을 변경 후 AGE 변수 80이상의 범죄율 평균 산출

In [2]:
df = pd.read_csv('boston.csv')

In [3]:
df

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     PTRATIO       B  LSTAT  MEDV  CAT. MEDV  
0       15.3  396.90   4.98  24.0          0  
1       17.8  396.90   9.14  21.6          0  
2       17.8  392.83   4.03  34.7          1  
3       18.7  394.63   2.94  33.4          1  
4       18.7  396.90   5.33  36.2          1  
..       ...     ...    ...   ...        ...  
501     21.0  391.99   9.67  22.4          0  
502     21.0  396.90   9.08  20.6          0  
503     21.0  396.90   5.64  23.9          0  
504     21.0  393.45   6.48  22.0          0  
505     21.0  396.90   7.88  11.9          0  

[506 rows x 15 columns]

In [4]:
sorted = df.sort_values(by='CRIM',ascending=False).copy()

In [5]:
sorted.head(10)

CRIM   ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD  TAX  \
380  88.9762  0.0   18.1     0  0.671  6.968   91.9  1.4165   24  666   
418  73.5341  0.0   18.1     0  0.679  5.957  100.0  1.8026   24  666   
405  67.9208  0.0   18.1     0  0.693  5.683  100.0  1.4254   24  666   
410  51.1358  0.0   18.1     0  0.597  5.757  100.0  1.4130   24  666   
414  45.7461  0.0   18.1     0  0.693  4.519  100.0  1.6582   24  666   
404  41.5292  0.0   18.1     0  0.693  5.531   85.4  1.6074   24  666   
398  38.3518  0.0   18.1     0  0.693  5.453  100.0  1.4896   24  666   
427  37.6619  0.0   18.1     0  0.679  6.202   78.7  1.8629   24  666   
413  28.6558  0.0   18.1     0  0.597  5.155  100.0  1.5894   24  666   
417  25.9406  0.0   18.1     0  0.679  5.304   89.1  1.6475   24  666   

     PTRATIO       B  LSTAT  MEDV  CAT. MEDV  
380     20.2  396.90  17.21  10.4          0  
418     20.2   16.45  20.62   8.8          0  
405     20.2  384.97  22.98   5.0          0  
410     20.2    2.60  10.11  15.0          0  
414     20.2   88.27  36.98   7.0          0  
404     20.2  329.46  27.38   8.5          0  
398     20.2  396.90  30.59   5.0          0  
427     20.2   18.82  14.52  10.9          0  
413     20.2  210.97  20.08  16.3          0  
417     20.2  127.36  26.64  10.4          0

In [6]:
sorted.iloc[9,0]

25.9406

In [7]:
sorted.iloc[:10,0] = sorted.iloc[9,0]

In [8]:
sorted.head(10)

CRIM   ZN  INDUS  CHAS    NOX     RM    AGE     DIS  RAD  TAX  \
380  25.9406  0.0   18.1     0  0.671  6.968   91.9  1.4165   24  666   
418  25.9406  0.0   18.1     0  0.679  5.957  100.0  1.8026   24  666   
405  25.9406  0.0   18.1     0  0.693  5.683  100.0  1.4254   24  666   
410  25.9406  0.0   18.1     0  0.597  5.757  100.0  1.4130   24  666   
414  25.9406  0.0   18.1     0  0.693  4.519  100.0  1.6582   24  666   
404  25.9406  0.0   18.1     0  0.693  5.531   85.4  1.6074   24  666   
398  25.9406  0.0   18.1     0  0.693  5.453  100.0  1.4896   24  666   
427  25.9406  0.0   18.1     0  0.679  6.202   78.7  1.8629   24  666   
413  25.9406  0.0   18.1     0  0.597  5.155  100.0  1.5894   24  666   
417  25.9406  0.0   18.1     0  0.679  5.304   89.1  1.6475   24  666   

     PTRATIO       B  LSTAT  MEDV  CAT. MEDV  
380     20.2  396.90  17.21  10.4          0  
418     20.2   16.45  20.62   8.8          0  
405     20.2  384.97  22.98   5.0          0  
410     20.2    2.60  10.11  15.0          0  
414     20.2   88.27  36.98   7.0          0  
404     20.2  329.46  27.38   8.5          0  
398     20.2  396.90  30.59   5.0          0  
427     20.2   18.82  14.52  10.9          0  
413     20.2  210.97  20.08  16.3          0  
417     20.2  127.36  26.64  10.4          0

In [9]:
sorted.loc[df.AGE>=80,'CRIM'].mean()

5.759386625

# 작업형 1-2
### 주어진 데이터 첫번째 행 부터 순서대로 80%까지의 데이터를 추출 후 'total_bedrooms' 변수의 결측값(NA)을 'total_bedrooms' 변수의 중앙값으로 대체하고 대체 전의 'total_bedrooms' 변수 표준편차값과 대체 후의 'total_bedrooms' 변수 표준편차 값 산출

In [31]:
df12 = pd.read_csv('housing.csv')

In [33]:
df12 = df12[:int(len(df12)*0.8)]

In [34]:
df12.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        159
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [35]:
bedroom_medi = df12['total_bedrooms'].median()
bedroom_medi

436.0

In [36]:
after = df12.fillna(bedroom_medi)
df12.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [37]:
after.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64

In [38]:
abs(after['total_bedrooms'].std() - df12['total_bedrooms'].std())

1.9751472916456692

# 작업형 1-3 이상치 처리 데이터 모르겠음
### 데이터의 특정컬럼의 이상치(이상치 기준 : 평균 + (표준편차 * 1.5) )를 찾아 이상치들의 합 산출 
데이터 : 뭔가 넥플릭스 데이터 같은 데이터임 (건수가 기억안남)

풀이 : 

평균 60, 표준편차 30, 표준편차 *1.5 = 45

105이상? 220 건 

19.5이하? 317 건

정답 : 약 32000? 

# 작업형 2